packages/core/src/services/chains/buildStep/index.ts

```diff
@@ -0,0 +1,93 @@
+import { Chain as PromptlChain } from 'promptl-ai'
+import { FinishReason } from 'ai'
+
+import { ChainStepResponse, StreamType } from '../../../constants'
+import { ChainStreamConsumer } from '../ChainStreamConsumer'
+import { ValidatedStep } from '../ChainValidator'
+import { runStep, StepProps } from '../run'
+import {
+  buildProviderLogDto,
+  saveOrPublishProviderLogs,
+} from '../ProviderProcessor/saveOrPublishProviderLogs'
+import { cacheChain } from '../chainCache'
+
+function hasToolCalls({
+  response,
+}: {
+  response: ChainStepResponse<StreamType>
+}) {
+  const type = response.streamType
+  if (type === 'object') return false
+
+  const toolCalls = response.toolCalls ?? []
+  return toolCalls.length > 0
+}
+
+export async function buildStepExecution({
+  streamConsumer,
+  baseResponse,
+  step,
+  stepProps,
+  providerLogProps: { streamType, finishReason, stepStartTime },
+}: {
+  streamConsumer: ChainStreamConsumer
+  baseResponse: ChainStepResponse<StreamType>
+  step: ValidatedStep
+  stepProps: StepProps
+  providerLogProps: {
+    streamType: StreamType
+    finishReason: FinishReason
+    stepStartTime: number
+  }
+}) {
+  const { chain } = stepProps
+  const workspace = stepProps.workspace
+  const documentLogUuid = stepProps.errorableUuid
+  const providerLog = await saveOrPublishProviderLogs({
+    workspace,
+    streamType,
+    finishReason,
+    data: buildProviderLogDto({
+      workspace,
+      provider: step.provider,
+      conversation: step.conversation,
+      source: stepProps.source,
+      errorableUuid: documentLogUuid,
+      stepStartTime,
+      response: baseResponse,
+    }),
+    // TODO: temp bugfix, shuold only save last one syncronously
+    saveSyncProviderLogs: true,
+  })
+
+  async function executeStep({
+    finalResponse,
+  }: {
+    finalResponse: ChainStepResponse<StreamType>
+  }): Promise<ChainStepResponse<StreamType>> {
+    const isPromptl = chain instanceof PromptlChain
+    const hasTools = isPromptl && hasToolCalls({ response: finalResponse })
+
+    if (hasTools) {
+      await cacheChain({ workspace, chain, documentLogUuid })
+    }
+
+    if (step.chainCompleted || hasTools) {
```

Comment:
I'm not sure about this. This is a breaking behaviour. 

To keep compatibility with existing code and also to ship this only for playground I think I'll change this to one of these 2 options:

1. Emit a new event `ChainStepWithToolCall` and stop the playground UI stream
2. Check on step completed event if there are toolCalls and stop the playground UI. 

All this file is a refactor to DRY code that was happening when:
1. Prompt run was cached
2. Prompt run was not cached.

But in terms of behaviour change I think this is the only change.